From 7e6abbc0ee5d6604fb3bfdf97c1f067e30a0b293 Mon Sep 17 00:00:00 2001
From: Robert Marko <robimarko@gmail.com>
Date: Tue, 8 Jun 2021 23:24:43 +0200
Subject: [PATCH 3/4] DMA: Fix NULL pointer exceptions

There are multiple instances that pass NULL instead
of device to DMA functions.
That is incorrect and will cause kernel NULL pointer
exceptions.

So, simply pass the device structure pointers.

Signed-off-by: Robert Marko <robimarko@gmail.com>
---
 nss_core.c     | 2 +-
 nss_coredump.c | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/nss_core.c b/nss_core.c
index 050da05..08d0121 100644
--- a/nss_core.c
+++ b/nss_core.c
@@ -1650,7 +1650,7 @@ static int32_t nss_core_handle_cause_queue(struct int_ctx_instance *int_ctx, uin
 		 *
 		 */
 		if (unlikely((buffer_type == N2H_BUFFER_CRYPTO_RESP))) {
-			dma_unmap_single(NULL, (desc->buffer + desc->payload_offs), desc->payload_len, DMA_FROM_DEVICE);
+			dma_unmap_single(nss_ctx->dev, (desc->buffer + desc->payload_offs), desc->payload_len, DMA_FROM_DEVICE);
 			goto consume;
 		}
 
diff --git a/nss_coredump.c b/nss_coredump.c
index c555707..faa6c4d 100644
--- a/nss_coredump.c
+++ b/nss_coredump.c
@@ -161,7 +161,7 @@ void nss_fw_coredump_notify(struct nss_ctx_instance *nss_own,
 		dma_addr = nss_own->meminfo_ctx.logbuffer_dma;
 	}
 
-	dma_sync_single_for_cpu(NULL, dma_addr, sizeof(struct nss_log_descriptor), DMA_FROM_DEVICE);
+	dma_sync_single_for_cpu(nss_own->dev, dma_addr, sizeof(struct nss_log_descriptor), DMA_FROM_DEVICE);
 
 	/*
 	 * If the current entry is smaller than or equal to the number of NSS_LOG_COREDUMP_LINE_NUM,
@@ -188,7 +188,7 @@ void nss_fw_coredump_notify(struct nss_ctx_instance *nss_own,
 
 		offset = (index * sizeof(struct nss_log_entry))
 			+ offsetof(struct nss_log_descriptor, log_ring_buffer);
-		dma_sync_single_for_cpu(NULL, dma_addr + offset,
+		dma_sync_single_for_cpu(nss_own->dev, dma_addr + offset,
 				sizeof(struct nss_log_entry), DMA_FROM_DEVICE);
 		nss_info_always("%px: %s\n", nss_own, nle_print->message);
 		nle_print++;
-- 
2.39.0

